hips/hip-0019.md

```diff
@@ -0,0 +1,120 @@
+---
+hip: "0019"
+title: "Adoption of CUE for Values Processing and Validation in Helm"
+authors: ["Jiří Altman <jiri.altman@usu.com>", "Jan von Löwenstein <jan.von.loewenstein@sap.com>"]
+created: "2024-08-07"
+type: "feature"
+status: "draft"
+---
+
+## Abstract
+
+This HIP proposes the integration of CUE (Configuration Unification Engine) as the primary tool for processing and validating Helm values files, replacing the existing JSON Schema support. As a superset of YAML, JSON, and JSON Schema, CUE extends existing capabilities while maintaining complete backward compatibility. This ensures that all current Helm charts and configurations remain fully supported, with the added benefits of enhanced validation, improved defaulting mechanisms, and easier schema management.
+
+## Motivation
```

Comment:
One significant benefit to utilizing CUE for values processing, is that CUE would allow "pre-processing" of input values (beyond validation).

A frequent pain-point of Helm charts is that values tend to be very "raw". It is difficult for values to not simply be the exact inputs to substitute in templates. As any business logic that expands or modifies the values needs to be written in gotemplate, either as a "named template" (function) or inline with the template logic. And gotemplate is both a terrible language to write configuration "business" logic. And the actual rendering of charts IMHO should contain as little business logic as possible (should be relegated as much as possible to dumb value substitution).

One motivation to me would be to enable mutation of the input values (which CUE excels at). In order to "expand" the user provided values into values suitable for the  (gotemplate) rendering phase.

This HIP could just limit CUE to validation only. But it might be worth considering if CUE could be used beyond validation?

Reply:
I think, you might be thinking about the above in this point? If so, I would recommend expanding that

> **Enhanced Features**: CUE natively supports advanced features such as defaulting, flexible referencing of values, and other configuration management capabilities critical for complex Helm deployments.